SU2_CFD/src/output/filewriter/CSTLFileWriter.cpp

```diff
@@ -0,0 +1,348 @@
+   partitions so we can set up buffers. The master node will handle
+   the writing of the STL file after gathering all of the data. ---*/
+
+  nLocalTria = dataSorter->GetnElem(TRIANGLE);
+  nLocalQuad = dataSorter->GetnElem(QUADRILATERAL);
+  nLocalTriaAll = nLocalTria + nLocalQuad*2; // Quad splitted into 2 tris
+
+  if (rank == MASTER_NODE) Buffer_Recv_nTriaAll = new unsigned long[nProcessor];
+
+  /*--- Communicate the maximum of local triangles on any process to each partition and the number of local vertices on each partition
+   to the master node with collective calls. ---*/
+
+  SU2_MPI::Allreduce(&nLocalTriaAll, &max_nLocalTriaAll, 1,
+                     MPI_UNSIGNED_LONG, MPI_MAX, MPI_COMM_WORLD);
+
+  max_nLocalCoords = max_nLocalTriaAll*3*3; /*--- 9 coordinates (3 per point) in a Triangle. ---*/
+
+  SU2_MPI::Gather(&nLocalTriaAll,       1, MPI_UNSIGNED_LONG,
+                  Buffer_Recv_nTriaAll, 1, MPI_UNSIGNED_LONG,
+                  MASTER_NODE, MPI_COMM_WORLD);
+
+  /*--- Allocate buffer for send/recv of the coordinate data. Only the master rank allocates buffers for the recv. ---*/
+  bufD_Send = new su2double[max_nLocalCoords](); /* Triangle has 3 Points with 3 coords each */
+  if (rank == MASTER_NODE)
+    bufD_Recv = new su2double[nProcessor*max_nLocalCoords];
+
+  /*--- Load send buffers with the local data on this rank. ---*/
+  /*--- Write triangle element coordinate data into send buffer---*/
+  unsigned long index = 0;
+  for (iElem = 0; iElem < nLocalTria; iElem++) {
+    for (iPoint = 0; iPoint < 3; iPoint++) {
+
+      global_node_number = dataSorter->GetElem_Connectivity(TRIANGLE, iElem, iPoint) - 1; // (var, GlobalPointindex)
+      unsigned long local_node_number = global_node_number - dataSorter->GetNodeBegin(rank);
+
+      for (iVar = 0; iVar < 3; iVar++){
+        if (dataSorter->FindProcessor(global_node_number) == rank) {
+          bufD_Send[index] = dataSorter->GetData(iVar, local_node_number);
+        } else {
+          bufD_Send[index] = GetHaloNodeValue(global_node_number, iVar);
+        }
+
+        index++;
+      }//iVar
+    }//iPoint
+  }//iElem
+
+  /*--Write quadrilateral element coordinate data into send buffer:
+    Each quad is split into two triangles after a fixed node order. ---*/
+  for (iElem = 0; iElem < nLocalQuad; iElem++) {
+    for (iPoint = 0; iPoint < Nodelist.size(); iPoint++) {
+      global_node_number = dataSorter->GetElem_Connectivity(QUADRILATERAL,iElem,Nodelist[iPoint]) - 1; // (var, GlobalPointindex)
+      unsigned long local_node_number = global_node_number - dataSorter->GetNodeBegin(rank);
+
+      for (iVar = 0; iVar < 3; iVar++){
+        if (dataSorter->FindProcessor(global_node_number) == rank) {
+          bufD_Send[index] = dataSorter->GetData(iVar, local_node_number);
+        } else {
+          bufD_Send[index] = GetHaloNodeValue(global_node_number, iVar);
+        }
+
+        index++;
+      }//iVar
+    }//iPoint

```

Comment:
I would make this loop a function and call it twice, once with TRIANGLE and a node list {0 1 2}, then with the QUADRILATERAL node list.

Reply:
You may also go over the node list with a range for (`for(auto node : Nodelist)`).
By the way there is a wee bit of a naming scheme mix going on, dataSorter Nodelist, local_node_number...

Reply:
Made a seperate function `StoreCoordData(enum GEO_TYPE elemType,...)` and used the range-for-loop.

I adopted `thisVariableStyle` for everything except the copied TecIO code, which I prefer to keep close to that implementation for comparability.